Storage/StorageTests/CoreData/CoreDataManagerTests.swift

```diff
@@ -70,28 +70,77 @@ final class CoreDataManagerTests: XCTestCase {
         XCTAssertEqual(viewContext.countObjects(ofType: ShippingLine.self), 0)
     }
 
-    func test_saving_derived_storage_while_resetting_CoreData_does_not_save_data() throws {
+    func test_performAndSave_executes_changes_in_background_then_updates_viewContext() throws {
         // Arrange
-        let manager = CoreDataManager(name: storageIdentifier, crashLogger: MockCrashLogger())
-        manager.reset()
+        let modelsInventory = try makeModelsInventory()
+        let manager = try makeManager(using: modelsInventory, deletingExistingStoreFiles: true)
         let viewContext = try XCTUnwrap(manager.viewStorage as? NSManagedObjectContext)
-        let derivedContext = try XCTUnwrap(manager.writerDerivedStorage as? NSManagedObjectContext)
+        XCTAssertEqual(viewContext.countObjects(ofType: Account.self), 0)
 
         // Action
-        waitForExpectation(count: 2) { expectation in
-            derivedContext.perform {
-                _ = derivedContext.insertNewObject(ofType: ShippingLine.self)
-            }
-            manager.saveDerivedType(derivedStorage: derivedContext, {
-                XCTAssertEqual(viewContext.countObjects(ofType: ShippingLine.self), 1)
+        waitForExpectation(count: 1) { expectation in
+            manager.performAndSave({ storage in
+                XCTAssertFalse(Thread.current.isMainThread)
+                self.insertAccount(to: storage)
+            }, completion: {
+                XCTAssertTrue(Thread.current.isMainThread)
                 expectation.fulfill()
-            })
-            manager.reset()
-            expectation.fulfill()
+            }, on: .main)
         }
 
         // Assert
-        XCTAssertEqual(viewContext.countObjects(ofType: ShippingLine.self), 0)
+        XCTAssertEqual(viewContext.countObjects(ofType: Account.self), 1)
+    }
+
+    func test_performAndSave_with_result_returns_correct_result_upon_success() throws {
+        // Arrange
+        let modelsInventory = try makeModelsInventory()
+        let manager = try makeManager(using: modelsInventory, deletingExistingStoreFiles: true)
+        let viewContext = try XCTUnwrap(manager.viewStorage as? NSManagedObjectContext)
```

Comment:
I know this a test, but there's a potential risk for a runtime error if `manager.viewStorage` cannot be unwrapped as an `NSManagedObjectContext`. For traceability, what if we update it to something like:
```
        guard let viewContext = manager.viewStorage as? NSManagedObjectContext else {
            XCTFail("Expected viewStorage to be an NSManagedObjectContext")
            return
        }
```

Reply:
I think it's redundant to do the guard check as `CoreDataManager` is supposed to use Core Data for the implentation. Throwing the error when there's an expected type of view storage should be preferrable as it's a serious issue for the class and should not be just a test failure. WDYT?